### DISCUSSION TITLE
Why does assigning to an unconstrained type parameter report nullable warning CS8603?

### DISCUSSION THREAD

**Discussion by FstTesla** [dotnet/roslyn#68175] [Q&A]
I have the following snippet:

```csharp
T Foo<T>(object? obj)
{
  return (T)obj;
}
```

**Why is the compiler reporting warnings [CS8603 and CS8600](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings?f1url=%3FappId%3Droslyn%26k%3Dk(CS8600)#possible-null-assigned-to-a-nonnullable-reference) on the `(T)obj` expression?**
As far as I understand, type parameter `T` has no constraint of being non-null. Indeed, it is perfectly legit to invoke the method in any of the ways below, two of which are nullable:

```csharp
int? fooOk1 = Foo<int?>(...);
int fooOk2 = Foo<int?>(...);
string? fooOk3 = Foo<string?>(...);
string fooOk4 = Foo<string >(...);
```

On the contrary, the following cases report a warning, as expected:

```csharp
int fooWarn1 = Foo<int?>(...);
string fooWarn2 = Foo<string?>(...);
```

---

Same goes for a *constrained nullable* type parameter like this:

```csharp
T Bar<T>(object? obj) where T : IConvertible?
{
  return (T)obj;
}

//int? barError1 = Bar<int?>(...); // Error CS0313, but it is fine
int barOk1 = Bar<int>(...);
string? barOk2 = Bar<string?>(...);
string barOk3 = Bar<string >(...);
int barWarn1 = Bar<int?>(...);
string barWarn2 = Bar<string?>(...);
```

---

Summing up, I think that assigning a nullable expression to a type parameter without non-null should not report warnings such as CS8603 or CS8600.

**Comment by RikkiGibson** [comment]
The warning is given for `(T)nullableObj` because of the possibility that `T` is a non-nullable type after substitution.

**Reply by FstTesla** [reply]
I agree of that possibility.

So, assuming that the logic before the assignment is null-sound, you say that it would be acceptable to suppress the warning with a null-forgiving operator (`(T)nullableObj!`).
Anyway, using the `!` here looks awkward to me, because _Its only role is to **change the null state of the expression to "not null"**, and to limit warnings given on its use._ (from [The null-forgiving operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/nullable-reference-types-specification#the-null-forgiving-operator), bold is mine). In other words, we are saying to the compiler that the operand is never null although it cannot infer it statically.
Instead, in this case the expression **can legally be null** if `T` is nullable, so the null-forgiving operator is just suppressing the warnings, therefore without the implied _operand is not null_ meaning.

I hope that my remark was clear.
